Sources/SwiftDiagnostics/DiagnosticDecorators/ANSIDiagnosticDecorator.swift

```diff
@@ -0,0 +1,248 @@
+  ///
+  /// ## Examples
+  ///
+  /// When `colorize` is set to true:
+  /// ```swift
+  /// let decorator = ANSIDiagnosticDecorator(colorize: true)
+  /// let decoratedMessage = decorator.decorateMessage("File not found", basedOnSeverity: .error)
+  /// // Output would be: "[1;31merror: [1;39mFile not found[0;0m"
+  /// ```
+  /// In this case, the prefix "error: " would be colorized, likely appearing in red.
+  ///
+  /// To achieve a similar colorized output manually in the console, you can use `printf` in bash:
+  /// ```bash
+  /// printf "\e[1;31merror: \e[1;39mFile not found\e[0;0m\n"
+  /// ```
+  /// This will print the "error: " prefix in red and the message "File not found" in the default text color.
+  ///
+  /// When `colorize` is set to false:
+  /// ```swift
+  /// let decorator = ANSIDiagnosticDecorator(colorize: false)
+  /// let decoratedMessage = decorator.decorateMessage("File not found", basedOnSeverity: .error)
+  /// // Output would be: "error: File not found"
+  /// ```
+  /// In this case, the prefix "error: " would appear as plain text, with no ANSI colorization applied.
+  public func decorateMessage(_ message: String, basedOnSeverity severity: DiagnosticSeverity) -> String {
+    let severityText: String
+    let severityAnnotation: ANSIAnnotation
+
+    switch severity {
+    case .error:
+      severityText = "error"
+      severityAnnotation = .errorText
+
+    case .warning:
+      severityText = "warning"
+      severityAnnotation = .warningText
+
+    case .note:
+      severityText = "note"
+      severityAnnotation = .noteText
+
+    case .remark:
+      severityText = "remark"
+      severityAnnotation = .remarkText
+    }
+
+    let prefix = colorizeIfRequested("\(severityText): ", usingAnnotation: severityAnnotation, resetAfterApplication: false)
+
+    return prefix + colorizeIfRequested(message, usingAnnotation: .diagnosticText)
+  }
+
+  /// Decorates the outline of a source code buffer, optionally applying ANSI cyan color colorization.
+  ///
+  /// - Parameter bufferOutline: The string representation of the source code buffer outline.
+  ///
+  /// - Returns: A string with optional ANSI cyan color colorization applied to the source code buffer outline.
+  public func decorateBufferOutline(_ bufferOutline: String) -> String {
+    colorizeIfRequested(bufferOutline, usingAnnotation: .bufferOutline)
+  }
+
+  /// Decorates a text segment intended for highlighting within a source code snippet. The method can also optionally
+  /// apply ANSI colorization for enhanced visual cues.
+  ///
+  /// - Parameter highlight: The text segment to be emphasized.
+  ///
+  /// - Returns: A tuple containing:
+  ///   - `highlightedSourceCode`: A string that represents the decorated version of the original source code snippet.
+  ///   - `additionalHighlightedLine`: always with nil.
```

Comment:
Why do we need to return `additionalHighlightedLine` if it’s always `nil`?

Reply:
It is `nil` in this implementation because the ANSI codes are able to underline lines directly. The assumption is that in future implementations it will contain characters like `~~~` or `---`.

Reply:
I would have a slight preference to only add `additionalHighlightedLine` once we actually add a diagnostic annotator that uses it, but I’m fine with keeping it as well.

Reply:
In the [Enhance diagnostic formatter to include trivia between highlight nodes](#2242) thread, you mentioned the idea of changing `assertSyntaxHighlightRanges` to actually underline the highlighted ranges, either via `DiagnosticDecorator` or a custom function. I think that's a really neat idea. Given that we plan to use `DiagnosticDecorator` in the next PR, I'd opt for keeping it as is for now.